enhancements/operator-conditions-lib.md

```diff
@@ -0,0 +1,116 @@
+
+## Design Details
+
+When an operator is installed in the cluster, OLM will create a `Condition` Custom Resource for the operator. The operator will be able to perform the following actions using the APIs provided in the library:
+
+1. Read the `Condition` CRD:
+
+The library will use controller-runtime's [client][cr_client] to access the resources present in the cluster. To fetch the resource, we will use the [`Get`][cr_get] method of the client, which requires the [`ObjectKey`][cr_objectkey] of type [`types.NamespacedName`][cr_namespacedname]. OLM will set the name of Condition CR as an [environment variable][conditions_followup_pr] while creating the resource. Namespace of the operator can be found from the associated [service account secret][accessing_api_from_pod] available in the operators' pod. The library will utilize both of these to obtain the CRD from cluster.
+
+2. Specify conditions in the CRD:
+
+The operators can introduce OLM-supported conditions which influence OLM's behavior or can also specify any custom (non-OLM supported) condition for communicating the status of the operator to users. An example Condition CR is given below:
+
+```yaml
+apiVersion: operators.coreos.com/v1
+kind: Condition
+metadata:
+  name: foo-operator-zdgen-asdf # RandomGen Name
+  namespace: operators
+status:
+   conditions:
+    type: Upgradeable # The name of the `foo` OLM Supported Condition.
+    status: "False"   # The operator is not ready to be upgraded.
+    reason: "migration"
+    message: "The operator is performing a migration."
+    lastTransitionTime: "2020-08-24T23:15:55Z"
+```
+
+The operator will be allowed to only modify the `status` field of condition CR. Operators can either add, delete or update the `status.conditions` array to include OLM-supported or custom conditions. The format and description of the fields present in the conditions can be referred [here][condition_type].
+
+The library will live in [`operator-lib`][operator_lib] repository and the definitions for Condition CRD will be present in [`operator-framwork/api`][api_repo]. The helper methods to work with [`conditions`][operator-lib-condition], which are present in operator-lib will be utilized to build APIs that will operate on the `Condition` CRD provided by OLM and perform necessary actions.
+
+#### Scenarios:
+
+1. Operator does not set/update any conditions that affect OLM:
+
+This is a no-op. In this case, OLM will default to its normal behavior.
+
+2. Operator tries to get/update condition CRD when it is not present in the cluster:
+
+In this case, the specific operation will return an error. The returned error can be wrapped, such that users can utilize the `errors.Is` method to compare the obtained error with specific values.
+
+**NOTE:**
+Operators cannot CREATE/DELETE the `Condition CR`. OLM will scope the operators' RBAC to only GET, LIST, UPDATE the Condition CR associated with their operator.
```

Comment:
We should make sure it also allows `PATCH`.

And should we document that the CRD will utilize the status subresource and operators will only be given these permissions on the status subresource (meaning no changing the spec or metadata)?

Reply:
Have documented it in Risks and Mitigations